lib/mongodb.nit

```diff
@@ -0,0 +1,701 @@
+	# var bson = new BSON.from_json_string(str)
+	# assert bson.to_s == str
+	# ~~~
+	new from_json_string(json_string: String) import String.to_cstring `{
+		bson_error_t error;
+		bson_t *bson;
+		bson = bson_new_from_json(String_to_cstring(json_string), -1, &error);
+		if(!bson) {
+	        printf ("%s\n", error.message);
+		}
+		return bson;
+	`}
+
+	redef fun to_s import NativeString.to_s `{
+		char *str = bson_as_json(recv, NULL);
+		return NativeString_to_s(str);
+	`}
+
+	# Returns a new JsonObject from `self`.
+	# ~~~
+	# var str = """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# var bson = new BSON.from_json_string(str)
+	# var json = bson.to_json
+	# assert json["age"] == 10
+	# assert json["name"] == "Rick"
+	# assert json["ELS"].as(JsonArray).is_empty
+	# ~~~
+	fun to_json: JsonObject do return to_s.parse_json.as(JsonObject)
+end
+
+# Introduces BSON services.
+redef class JsonObject
+
+	# Init `self` from a BSON object.
+	#
+	# ~~~
+	# var str = """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# var bson = new BSON.from_json_string(str)
+	# var json = new JsonObject.from_bson(bson)
+	# assert json["age"] == 10
+	# assert json["name"] == "Rick"
+	# assert json["ELS"].as(JsonArray).is_empty
+	# ~~~
+	init from_bson(bson: BSON) do
+		for k, v in bson.to_json do self[k] = v
+	end
+
+	# Returns a new BSON object from `self`.
+	#
+	# ~~~
+	# var obj = new JsonObject
+	# obj["age"] = 10
+	# obj["name"] = "Rick"
+	# obj["ELS"] = new JsonArray
+	# var bson = obj.to_bson
+	# assert bson.to_s == """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# ~~~
+	fun to_bson: BSON do return new BSON.from_json(self)
+end
+
+# An error returned by the mongoc client.
+#
+# Within the client, if a method returns `false` or `null` it's more likely that
+# an error occured during the execution.
+# You can see the last error occured with `Sys::last_mongoc_error`.
+#
+# Errors are formated in BSON.
+extern class BSONError `{ bson_error_t * `}
+
+	# Logical domain within a library that created the error.
+	fun domain: Int `{ return recv->domain; `}
+
+	# Domain specific error code.
+	fun code: Int `{ return recv->code; `}
+
+	# Human readable error message.
+	fun message: String import NativeString.to_s `{
+		return NativeString_to_s(recv->message);
+	`}
+
+	redef fun to_s do return "{message} (code: {code})"
+end
+
+redef class Sys
+	# Last error raised by `monogdb::MongoClient`.
+	var last_mongoc_error: nullable BSONError = null
+end
+
+# The MongoClient is used to connect to the mongo server and send queries.
+#
+# Usage:
+#
+# ~~~
+# var client = new MongoClient("mongodb://localhost:27017/")
+# client.close
+# ~~~
+extern class MongoClient `{ mongoc_client_t * `}
+
+	# Connect to `uri`.
+	new(uri: String) import String.to_cstring `{
+		mongoc_init();
+		return mongoc_client_new(String_to_cstring(uri));
+	`}
+
+	# Close the connexion between the client and MongoDb.
+	#
+	# FIXME: how to handle destructors?
```

Comment:
To handle the destroy, one idea to consider is to abstract this class and have the abstraction `super Finalizable`

It can be a good way to handle errors too, since you could add a flag specifying that the connection was closed, and any operation on `self` could NOP instead of crashing that way.